pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -382,6 +408,39 @@ func (d *ServiceEntryStore) GetIstioServiceAccounts(*model.Service, []int) []str
 	return nil
 }
 
+func servicesDiff(os []*model.Service, ns []*model.Service) ([]*model.Service, []*model.Service, []*model.Service, []*model.Service) {
+	var (
+		added, deleted, updated, unchanged []*model.Service
+	)
+
+	oldServiceHosts := make(map[string]*model.Service, len(os))
+	newServiceHosts := make(map[string]*model.Service, len(ns))
+	for _, s := range os {
+		oldServiceHosts[string(s.Hostname)] = s
+	}
+	for _, s := range ns {
+		newServiceHosts[string(s.Hostname)] = s
+	}
+
+	for name, oldSvc := range oldServiceHosts {
+		newSvc, f := newServiceHosts[name]
+		if !f {
+			deleted = append(deleted, oldSvc)
+		} else if !reflect.DeepEqual(oldSvc, newSvc) {
+			updated = append(updated, newSvc)
+		} else {
+			unchanged = append(unchanged, newSvc)
+		}
+	}
+	for name, newSvc := range newServiceHosts {
+		if _, f := oldServiceHosts[name]; !f {
+			added = append(added, newSvc)
+		}
+	}
+
+	return added, deleted, updated, unchanged
+}
+
 // This method compares if services have changed, that needs full push.
 func servicesChanged(os []*model.Service, ns []*model.Service) bool {
```

Comment:
dont you want to remove this function?

Reply:
OK, I didn't see that's the only usage of this function.

Reply:
As john mentioned above, I'll separate the ServiceEntry optimization to another PR and thus the function will still be used and preserved until that.